server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -69,9 +69,8 @@ public ParsedDocument parseDocument(SourceToParse source, MappingLookup mappingL
         } catch (Exception e) {
             throw wrapInMapperParsingException(source, e);
         }
-        String remainingPath = context.path().pathAsText("");
-        if (remainingPath.isEmpty() == false) {
-            throwOnLeftoverPathElements(remainingPath);
+        if (context.path.atRoot() == false) {
+            throwOnLeftoverPathElements(context.path);
```

Comment:
Since the exception is only thrown here, can we inline it for readability?
Also the case where this exception is thrown seems untested, would it be possible to add a check to DocumentParserTests?

Reply:
Hmm I think it might be impossible to get to this code?
I can only see getting here via broken JSON, but the JSON (or other type) parsers will throw on malformed input.
So it seems impossible to ever get through parsing any document without the appropriate number of closing brackets?

I the exception is thrown in a separate method as a bit of a questionable JIT optimization. But now that I think about the above ... can't we just remove this code?

Reply:
Sure, let me take another closer look to understand your reasoning, all in favor of removing it if we don't even test it.

Reply:
I took another look, the only way I see this exception can be thrown here is if there were calls to path.add() somewhere inside the parsing logic that wouldn't be correctly followed by a path.remove() operation. While that is certainly possible due to a programming error, that sounds much better suited to be a test or an assertion at this point than an exception. 
I would opt for make this an assertion to communicate the invariant we have here and potentially catch errors in tests where we have assertions enabled.